File: OBI/2021/M2/poligrama.py
ALPHABET_SIZE=26

def anagram(str1:str,str2:str,length:int)->int:
    #print(str1,str2)
    letters_signature=[0 for i in range(ALPHABET_SIZE)]
    for i in range(length):
        letters_signature[ord(str1[i])-ord('a')]+=1
        letters_signature[ord(str2[i])-ord('a')]-=1
    #print(letters_signature)
    for i in range(ALPHABET_SIZE):
        if letters_signature[i]:
            return 0
    return 1
        
def main()->None:
    x=int(input())
    string=input()

    for i in range(1,x):
        if x%i:
            continue
        valid=1
        for j in range(i,x,i):
            sub_str=string[j:x]
            if not anagram(string,sub_str,i):
                valid=0
                break
        if valid:
            print(string[:i])
            return
    print('*')
    return

if __name__=='__main__':
    main()

File: OBI/2021/M2/poligrama.cpp
#include<iostream>
#include<cstring>

#define ALPHABET_SIZE 26

int anagram(const char*str1,const char*str2,size_t str2_size){
    //printf("%s %s\n",str1,str2);
    int letters_signature[ALPHABET_SIZE]={0};
    for(size_t i=0;i<str2_size;++i){
        ++letters_signature[str1[i]-'a'];
        --letters_signature[str2[i]-'a'];
    }
    for(size_t i=0;i<ALPHABET_SIZE;++i)
        if(letters_signature[i])
            return 0;
    return 1;
}
//
int main(){
    size_t str_size;
    std::scanf("%zu",&str_size);
    char*str=(char*)malloc(sizeof(char)*(str_size+5));
    std::scanf("%s",str);

    for(size_t root_length=1;root_length<=str_size/2;++root_length){
        if(str_size%root_length)
            continue;
        int valid=1;
        for(size_t i=root_length;i<str_size;i+=root_length){
            if(!anagram(str,str+i,root_length)){
                valid=0;
                break;
            }
        }
        if(valid){
            printf("%s\n",strndup(str,root_length));
            return 0;
        }
    }

    printf("*\n");
    return 0;
}
